File: membrain-node/tests/test_agent.mjs
/**
 * Tests for the Membrain agent framework (Node.js).
 *
 * Core planner/executor/agent logic is exercised with deterministic in-process
 * LLM callables. A single smoke test runs against the real OpenAI API when
 * OPENAI_API_KEY is set.
 */

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { randomUUID } from "node:crypto";

import {
  CasePromptBuilder,
  MembrainAgent,
  MembrainClient,
  MembrainExecutor,
  MembrainPlanner,
  NonParametricRetriever,
} from "../dist/index.js";

function loadApiKey() {
  if (process.env.OPENAI_API_KEY) return process.env.OPENAI_API_KEY;
  const envPath = join(import.meta.dirname, "..", "..", ".env");
  if (!existsSync(envPath)) return null;
  for (const line of readFileSync(envPath, "utf-8").split("\n")) {
    if (line.startsWith("OPENAI_API_KEY=")) {
      return line.split("=")[1].trim();
    }
  }
  return null;
}

function makeTmpPath() {
  return join(tmpdir(), `membrain_agent_${randomUUID()}`);
}

function planJson(steps) {
  const payload = {
    plan: steps.map((description, index) => ({ id: index + 1, description })),
  };
  return JSON.stringify(payload);
}

class EmptyRetriever extends NonParametricRetriever {
  constructor() {
    super(null);
  }
  retrieve() {
    return { positive_cases: [], negative_cases: [], duration_ms: 0 };
  }
}

describe("MembrainPlanner", () => {
  it("returns a structured plan", async () => {
    const calls = [];
    const llm = async (messages) => {
      calls.push(messages);
      return planJson(["Gather requirements", "Write code", "Deploy"]);
    };

    const planner = new MembrainPlanner(
      new EmptyRetriever(),
      new CasePromptBuilder(),
      llm,
    );
    const plan = await planner.plan("Ship a new microservice");
    assert.deepEqual(
      plan.steps.map((step) => step.description),
      ["Gather requirements", "Write code", "Deploy"],
    );
    assert.equal(plan.steps[0].id, 1);
    assert.equal(calls[0][0].role, "system");
    assert.match(calls[0][0].content, /task planner/i);
  });

  it("strips markdown fences", async () => {
    const llm = async () => "```json\n" + planJson(["Only step"]) + "\n```";
    const planner = new MembrainPlanner(
      new EmptyRetriever(),
      new CasePromptBuilder(),
      llm,
    );
    const plan = await planner.plan("Query");
    assert.equal(plan.steps.length, 1);
    assert.equal(plan.steps[0].description, "Only step");
  });

  it("falls back to raw text on invalid JSON", async () => {
    const llm = async () => "not-valid-json";
    const planner = new MembrainPlanner(
      new EmptyRetriever(),
      new CasePromptBuilder(),
      llm,
    );
    const plan = await planner.plan("Query");
    assert.equal(plan.steps.length, 1);
    assert.equal(plan.steps[0].description, "not-valid-json");
  });
});

describe("MembrainExecutor", () => {
  it("dispatches a registered tool", async () => {
    const llm = async () => "TOOL_CALL: echo('hello world')";
    const executor = new MembrainExecutor(llm);
    executor.registerTool("echo", (value) => `echoed: ${value}`);
    const result = await executor.execute({ id: 1, description: "Echo" });
    assert.equal(result.success, true);
    assert.equal(result.output, "echoed: hello world");
    assert.deepEqual(executor.availableTools, ["echo"]);
  });

  it("passes through plain responses", async () => {
    const llm = async () => "direct answer";
    const executor = new MembrainExecutor(llm);
    const result = await executor.execute({ id: 1, description: "Task" });
    assert.equal(result.output, "direct answer");
    assert.equal(result.success, true);
  });

  it("captures LLM failures", async () => {
    const llm = async () => {
      throw new Error("LLM upstream failure");
    };
    const executor = new MembrainExecutor(llm);
    const result = await executor.execute({ id: 1, description: "Task" });
    assert.equal(result.success, false);
    assert.match(result.error, /LLM upstream failure/);
  });

  it("reports unknown tool names", async () => {
    const llm = async () => "TOOL_CALL: missing()";
    const executor = new MembrainExecutor(llm);
    const result = await executor.execute({ id: 1, description: "Task" });
    assert.equal(result.success, true);
    assert.match(result.output, /Unknown tool: missing/);
  });
});

describe("MembrainAgent", () => {
  it("runs plan-execute end-to-end", async () => {
    const client = new MembrainClient({ storage: { path: makeTmpPath() } });
    try {
      let count = 0;
      const llm = async (messages) => {
        count += 1;
        const system = messages[0]?.content ?? "";
        if (system.toLowerCase().includes("task planner")) {
          return planJson(["Do the task"]);
        }
        return "completed";
      };
      const agent = new MembrainAgent(client, llm);
      const output = await agent.run("Diagnose the production outage");
      assert.equal(output, "completed");
      assert.ok(count >= 2);
    } finally {
      client.close();
    }
  });

  it("stops at max cycles when steps fail", async () => {
    const client = new MembrainClient({ storage: { path: makeTmpPath() } });
    try {
      const llm = async (messages) => {
        const system = messages[0]?.content ?? "";
        if (system.toLowerCase().includes("task planner")) {
          return planJson(["Will fail"]);
        }
        throw new Error("step blew up");
      };
      const agent = new MembrainAgent(client, llm, { maxCycles: 2 });
      const output = await agent.run("Impossible task");
      assert.equal(output, "");
    } finally {
      client.close();
    }
  });
});

describe("MembrainAgent live OpenAI", { skip: !loadApiKey() }, () => {
  it("produces a non-empty answer against gpt-4o-mini", async () => {
    const { default: OpenAI } = await import("openai").catch(() => ({
      default: null,
    }));
    if (!OpenAI) return;

    const client = new MembrainClient({ storage: { path: makeTmpPath() } });
    try {
      const openai = new OpenAI({ apiKey: loadApiKey() });
      const llm = async (messages) => {
        const response = await openai.chat.completions.create({
          model: "gpt-4o-mini",
          messages,
          temperature: 0,
        });
        return response.choices[0].message.content ?? "";
      };
      const agent = new MembrainAgent(client, llm, { maxCycles: 1 });
      const output = await agent.run(
        "Summarise the three steps to deploy a node web service",
      );
      assert.ok(output.length > 10);
    } finally {
      client.close();
    }
  });
});

File: membrain-py/tests/test_agent.py
"""Tests for the Membrain agent framework.

The planner, executor, and top-level agent are exercised against the real
MembrainClient (no FFI mocks). The LLM layer is a user-supplied callable, so
tests use deterministic in-process callables for core logic and reserve the
live OpenAI path behind `OPENAI_API_KEY` for a single smoke test.
"""

from __future__ import annotations

import json
import os

import pytest

from membrain import MembrainClient
from membrain.agent import (
    MembrainAgent,
    MembrainExecutor,
    MembrainPlanner,
    TaskPlan,
    TaskStep,
)
from membrain.cbr import CasePromptBuilder, NonParametricRetriever


@pytest.fixture
def client(tmp_path):
    config = {"storage": {"backend": "memscaledb", "path": str(tmp_path / "db")}}
    instance = MembrainClient(config=config)
    try:
        yield instance
    finally:
        instance.close()


def _plan_json(steps: list[str]) -> str:
    payload = {
        "plan": [
            {"id": index + 1, "description": text}
            for index, text in enumerate(steps)
        ]
    }
    return json.dumps(payload)


def test_planner_returns_structured_plan(client):
    recorded: list[list[dict[str, str]]] = []

    def llm(messages: list[dict[str, str]]) -> str:
        recorded.append(messages)
        return _plan_json(["Gather requirements", "Write code", "Deploy"])

    planner = MembrainPlanner(
        retriever=NonParametricRetriever(client),
        prompt_builder=CasePromptBuilder(),
        llm_callable=llm,
    )

    plan = planner.plan("Ship a new microservice")

    assert isinstance(plan, TaskPlan)
    assert [step.description for step in plan.steps] == [
        "Gather requirements",
        "Write code",
        "Deploy",
    ]
    assert plan.steps[0].id == 1
    assert recorded[0][0]["role"] == "system"
    assert "task planner" in recorded[0][0]["content"].lower()


def test_planner_strips_code_fences():
    def llm(_: list[dict[str, str]]) -> str:
        return "```json\n" + _plan_json(["Step one"]) + "\n```"

    planner = MembrainPlanner(
        retriever=_EmptyRetriever(),
        prompt_builder=CasePromptBuilder(),
        llm_callable=llm,
    )
    plan = planner.plan("Query")
    assert len(plan.steps) == 1
    assert plan.steps[0].description == "Step one"


def test_planner_fallback_on_bad_json():
    def llm(_: list[dict[str, str]]) -> str:
        return "not-valid-json"

    planner = MembrainPlanner(
        retriever=_EmptyRetriever(),
        prompt_builder=CasePromptBuilder(),
        llm_callable=llm,
    )
    plan = planner.plan("Query")
    assert len(plan.steps) == 1
    assert plan.steps[0].description == "not-valid-json"


def test_executor_dispatches_registered_tool():
    def llm(_: list[dict[str, str]]) -> str:
        return "TOOL_CALL: echo('hello world')"

    executor = MembrainExecutor(llm_callable=llm)
    executor.register_tool("echo", lambda value: f"echoed: {value}")
    result = executor.execute(TaskStep(id=1, description="Echo 'hello world'"))

    assert result.success is True
    assert result.output == "echoed: hello world"
    assert "echo" in executor.available_tools


def test_executor_passes_through_plain_response():
    def llm(_: list[dict[str, str]]) -> str:
        return "direct answer"

    executor = MembrainExecutor(llm_callable=llm)
    result = executor.execute(TaskStep(id=1, description="Task"))
    assert result.output == "direct answer"
    assert result.success is True


def test_executor_captures_failures():
    def llm(_: list[dict[str, str]]) -> str:
        raise RuntimeError("LLM upstream failure")

    executor = MembrainExecutor(llm_callable=llm)
    result = executor.execute(TaskStep(id=1, description="Task"))
    assert result.success is False
    assert result.error is not None
    assert "LLM upstream failure" in result.error


def test_agent_run_end_to_end(client):
    call_count = {"value": 0}

    def llm(messages: list[dict[str, str]]) -> str:
        call_count["value"] += 1
        system = messages[0]["content"] if messages else ""
        if "task planner" in system.lower():
            return _plan_json(["Do the task"])
        return "completed"

    agent = MembrainAgent(client=client, llm_callable=llm)
    output = agent.run("Diagnose the production outage")

    assert output == "completed"
    assert call_count["value"] >= 2


@pytest.mark.skipif(
    not os.environ.get("OPENAI_API_KEY"),
    reason="OPENAI_API_KEY not set",
)
def test_agent_run_with_live_openai(client):
    from openai import AuthenticationError, OpenAI

    openai_client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])

    def llm(messages: list[dict[str, str]]) -> str:
        response = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=messages,
            temperature=0,
        )
        return response.choices[0].message.content or ""

    agent = MembrainAgent(
        client=client, llm_callable=llm, max_cycles=1
    )
    try:
        output = agent.run(
            "Summarise the three steps to deploy a python web service"
        )
    except AuthenticationError:
        pytest.skip("OpenAI credentials rejected (401)")
    assert output
    assert len(output) > 10


class _EmptyRetriever(NonParametricRetriever):
    def __init__(self) -> None:
        pass

    def retrieve(self, query, limit=5, positive_reward_threshold=0.5):
        from membrain.types import CaseSearchResults

        return CaseSearchResults(
            positive_cases=[], negative_cases=[], duration_ms=0.0
        )
